src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala

```diff
@@ -311,4 +315,127 @@ object LoadBalancer {
+                                       math.pow(leftBlock.sphereCenterY - rightBlock.sphereCenterY, 2) +
+                                       math.pow(leftBlock.sphereCenterZ - rightBlock.sphereCenterZ, 2))
+        sphereDistance < (leftBlock.sphereRadius + rightBlock.sphereRadius)
+      }
+
+      /*
+       * We also don't want to consider any right blocks that are separated
+       * from the current left block by a second joint, indicated by the
+       * presence of equal and opposite face normal vectors
+       */
+      val relevantLeftFaces = excludeProcessorFace(leftBlock.faces, commonJoint)
+      val rightBlocksRelevantFaces = sphereCandidates.map { rightBlock =>
+        (rightBlock, excludeProcessorFace(rightBlock.faces, commonJoint))
+      }
+
+      val mateCandidates = rightBlocksRelevantFaces.filter { case (rightBlock, relevantRightFaces) =>
+        !relevantLeftFaces.exists(leftFace => relevantRightFaces.exists { rightFace =>
+          math.abs(leftFace.a + rightFace.a) <= NumericUtils.EPSILON &&
+          math.abs(leftFace.b + rightFace.b) <= NumericUtils.EPSILON &&
+          math.abs(leftFace.c + rightFace.c) <= NumericUtils.EPSILON &&
+          math.abs(leftFace.d + rightFace.d) <= NumericUtils.EPSILON
+        })
+      }
+
+      // Use a 'view' to avoid unnecessary LP computation
+      val candidateBlocks = mateCandidates.view.map { case (rightBlock, relevantRightFaces) =>
+        val globalOrigin = Array(0.0, 0.0, 0.0)
+        // No need to update newFaces since the inputs are already referenced to global origin
+        val newBlock = Block(globalOrigin, relevantLeftFaces ++ relevantRightFaces)
+        (rightBlock, Block(globalOrigin, newBlock.nonRedundantFaces))
+      }
+
+      val matchOpt = candidateBlocks.find(_._2.faces.nonEmpty)
+      if (matchOpt.isDefined) {
+        val (rightMate, mergedBlock) = matchOpt.get
+        remainingRightBlocks.remove(rightMate)
+        if (mergedBlock.faces.exists(_.isProcessorFace)) {
+          orphanBlocks.add(mergedBlock)
+        } else {
+          mergedBlocks.add(mergedBlock)
+        }
+      } else {
+        orphanBlocks.add(leftBlock)
+      }
+    }
+
+    orphanBlocks ++= remainingRightBlocks
+    (mergedBlocks.toSeq, orphanBlocks.toSeq)
+  }
+
+  /**
+    * Given a collection of processor blocks (blocks containing at least one
+    * processor face), merge any blocks that share a common processor face.
+    *
+    * @param processorBlocks A Seq of blocks, each of which must contain at
+    *                        least one processor face.
+    * @return A Seq of merged blocks. None of them should contain a processor
+    *         face after the merging process.
+    */
+  def mergeProcessorBlocks(processorBlocks: Seq[Block]): Seq[Block] = {
+    val globalOrigin = Array(0.0, 0.0, 0.0)
+    var orphanBlocks = processorBlocks.map { block =>
+      // TODO: Requires 4 decimal places to pass unit tests. We should look at this.
```

Comment:
Yeah, I ran into this same problem with the first iteration of merging blocks. Not sure how we can limit this, but that's why there were all those different tolerances floating around. Do you have any suggestions for how we can deal with this? Part of me thinks that matching to the 4th decimal place is okay, but at the same time it seems that we should be able to control this better.
Most likely, the majority of the error is coming from the LP solution. We take the solution to the LP for intersection as the new origin for the blocks. Since the solution to the LP is iterative, the solution can vary depending on which block you're looking at.
Maybe we can look at the options in `lpsolve`, perhaps there's a way to specify tighter tolerance on the LP solution.